File: ruby/sieve/sieve.rb
class Sieve
  def initialize(max)
    @max = max
  end

  def primes
    res = (2..@max).to_a
    sqrt_of_max = Math.sqrt @max
    res.each do |prime|
      return res if prime > sqrt_of_max
      res.delete_if do |num|
        d, m = num.divmod prime
        d != 1 && m == 0
      end
    end
  end
end

File: go/sieve/sieve.go
package sieve

func Sieve(upto int) (res []int) {
	nums := twoToN(upto)

	for n := 2; n <= upto; n++ {
		if !nums[n] {
			continue
		}
		for i := 1; i*n <= upto; i++ {

			nums[i*n] = false
		}
		res = append(res, n)
	}
	return
}

func twoToN(n int) (res map[int]bool) {
	res = make(map[int]bool)
	for i := 2; i <= n; i++ {
		res[i] = true
	}
	return
}
